src/streamer/jamstreamer.cpp

```diff
@@ -0,0 +1,35 @@
+#include "jamstreamer.h"
+
+using namespace streamer;
+
+// --- CONSTRUCTOR ---
+CJamStreamer::CJamStreamer() {}
+
+// --- PROCESS ---
+// put the received pcm data into the pipe to ffmpeg
+void CJamStreamer::process( int iServerFrameSizeSamples, CVector<int16_t> data ) {
+    if (pcm == 0) // thanks to hselasky
+    {
+        pcm = new int16_t [2 * iServerFrameSizeSamples];
+    }
+    for ( int i = 0; i < ( 2 * iServerFrameSizeSamples ); i++ )
+    {
+        pcm[i] = data[i];
+    }
+    fwrite(pcm, 2, ( 2 * iServerFrameSizeSamples ), pipeout);
+}
+
+void CJamStreamer::Init( const QString strStreamDest ) {
+    this->strStreamDest = strStreamDest;
+}
+
+void CJamStreamer::OnStarted() {
+    // create a pipe to ffmpeg called "pipeout" to being able to put out the pcm data to it
+    pcm = 0; // it is necessary to initialise pcm
+    QString command = "ffmpeg -y -f s16le -ar 48000 -ac 2 -i - " + strStreamDest;
```

Comment:
It might be worth considering hardening the ffmpeg call as it executes a command and includes a variable in the invocation.  fail to find a `popen()` alternative which takes an array of arguments instead of a string (i.e. like `exec*()` to `system()`). Unless I'm missing something, the only remaining hardening measure would be proper quoting + escaping of the variable. One might argue that the variable is currently supposed to come from a trusted context (command line), but things may change in unexpected ways sometimes. Improving this would also ensure that special chars will not break things even in well-intentioned environments.

Reply:
Perhaps use [`QProcess`](https://doc.qt.io/qt-5/qprocess.html)?

Reply:
> Perhaps use [`QProcess`](https://doc.qt.io/qt-5/qprocess.html)?

Good idea! This should solve the potential security impact and might also get us cross-platform support for free. :)

Reply:
Hm.  I'm thinking you're going to want to _know_ the path the ffmpeg binary.  Ideally from either a setting or from a command line option.  Better not just to launch whatever executable happens to get picked up from "ffmpeg" getting run.

Reply:
> the only remaining hardening measure would be proper quoting + escaping of the variable

How would I go about and do this? At the moment I just read the output part of ffmpeg from the commandline (which can be anything that ffmpeg understands like "-f mp3 icecast://source:<password>:<port>/<mountpoint>" or something completely different. I don't know in advance how many arguments will be given in different usecases)

> Perhaps use QProcess?

Thanks for the advice! I looked into it and can't see how I could alter my code (I'm not a programmer ;) to do the same as before using QProcess. Any hints on what I would be looking for?

> Better not just to launch whatever executable happens to get picked up from "ffmpeg" getting run.

I've thought about that before and come to the conclusion that it won't matter for security reasons since you can call anything named "ffmpeg". Allowing the user to specify which program to call doesn't make it more secure in my opinion, since you'd be able to call anything you want. It'd only be useful if you wanted to call a certain build of ffmpeg besides the one already installed system-wide one, but I couldn't think of a usecase where you would want to do that.

Reply:
> I looked into it and can't see how I could alter my code (I'm not a programmer ;) to do the same as before using QProcess.

I haven't done this before, but it should work like: `FILE* pipeout` is replaced by `QProcess qproc`, the `popen` call would be replaced by

```
QStringList argumentList = { "ffmpeg", "-y", "-f", "s16le", "-ar", "48000", "-ac", "2", "-i", "-", strStreamDest };
qproc.start ( "ffmpeg", argumentList, QIODevice::WriteOnly ); // Note that program name is also repeated as first argument
```

Then the `fwrite(..., pipeout)` would be replaced by `qproc.write(...)`. And `pclose(pipeout)` becomes `qproc.closeWriteChannel()`

Reply:
almost there :) the QProcess.write call still complains about the given arguments.
could you try compiling yourself and see what the output is? I've tried a few things and it only made it worse...

Reply:
Oh yeah, `QProcess.write` takes `const char*` rather than `const void*` so some more casting is needed:

        qproc.write ( reinterpret_cast<const char*> ( &data[0] ), sizeof (int16_t) * ( 2 * iServerFrameSizeSamples ) );

Reply:
thanks! this compiles now but crashes the server with this message:
```
QObject: Cannot create children for a parent that is in a different thread.
QSocketNotifier: Socket notifiers cannot be enabled or disabled from another thread
```
i'm creating a thread and connecting it with the signals in server.cpp and i'm totally new to all this.
sorry for asking too many questions :/

Reply:
Oh, hmm. I'm not really familiar with how threading works for Qt either, but maybe the fix is to make `qproc` into a pointer, so that the object can be constructed only after the `moveToThread` finishes? That way it would be constructed and used in the same thread, I think.

Reply:
making it a pointer gets rid of the error messages but the server still just silently segfaults. we're close, i guess ;)
could you or anybody else ( maybe @hoffie , @pljones ) give this code a test run to see what might be happening there?
you can launch the server with:
```
--streamto "-f mp3 output.mp3"
```

Reply:
Well, it doesn't crash for me, but I'm not able to get any output either with the `QProcess*` version. It looks like `ffmpeg` never even got started, although I didn't manage to catch any `QProcess::FailedToStart` error. Maybe we should just stick with the `popen` version...
https://github.com/dingodoppelt/jamulus/compare/streamer2...npostavs:streamer2.np.qproc

Reply:
> Well, it doesn't crash for me

It does for me only when I connect a client but yes, ffmpeg never gets started it seems.
> Maybe we should just stick with the popen version...

I think so, too

Reply:
I'm not seeing a segfault either, but I have found three problems which I have fixed in a PR to @npostavs' branch against this PR. :)

https://github.com/npostavs/jamulus/pull/1

Reply:
Awesome! Thanks a lot for your work :)
Would that mean I can get rid of the conditional compile for win32 as well? I can't check since I have no windows machine here.

Reply:
> Would that mean I can get rid of the conditional compile for win32 as well?

I think so, yes, but one should test.

> I can't check since I have no windows machine here.

Neither can I (Linux-only here). A first smoke test would probably be removing the ifdefs. CodeQL should probably catch if compiling works. Someone should still run a real test with ffmpeg if possible.

Reply:
It doesn't compile on windows whatsoever. I've put the conditional compile back in